module/zfs/vdev_disk.c

```diff
@@ -638,13 +639,28 @@ __vdev_disk_physio(struct block_device *bdev, zio_t *zio,
 		}
 
 		/* bio_alloc() with __GFP_WAIT never returns NULL */
-		dr->dr_bio[i] = bio_alloc(GFP_NOIO,
-		    MIN(abd_nr_pages_off(zio->io_abd, bio_size, abd_offset),
-		    BIO_MAX_PAGES));
+		nr_iovecs = MIN(abd_nr_pages_off(zio->io_abd, bio_size,
+		    abd_offset), BIO_MAX_PAGES);
+		dr->dr_bio[i] = bio_alloc(GFP_NOIO, nr_iovecs);
 		if (unlikely(dr->dr_bio[i] == NULL)) {
```

Comment:
What if we instead did something a little less disruptive to the overall flow of the code.  This was clearly a brief regression in an -rc kernel which was quickly fixed.  Let's simply sanity check the bi_max_vecs in the existing NULL check, like this.  It'll be reported as ENOMEM instead of EIO but that seems reasonable since we didn't get what we asked for.

```diff
@@ -641,7 +641,8 @@ retry:
                dr->dr_bio[i] = bio_alloc(GFP_NOIO,
                    MIN(abd_nr_pages_off(zio->io_abd, bio_size, abd_offset),
                    BIO_MAX_PAGES));
-               if (unlikely(dr->dr_bio[i] == NULL)) {
+               if (unlikely(dr->dr_bio[i] == NULL ||
+                   (unsigned int)dr->dr_bio[i]->bi_max_vecs != nr_iovecs)) {
                        vdev_disk_dio_free(dr);
                        return (SET_ERROR(ENOMEM));
                }
```

Reply:
ok.

Btw I don't think anything is fixed at least in blk layer. I probably just happened to see it in -rc1 in ZoL (not other fs). I've briefly looked through git log, but didn't see anything implied fix for this.

Reply:
Updated

Reply:
> but didn't see anything implied fix for this.

That's a little troubling, but as I understand it you're no longer able to reproduce this in latter -rc's?  The updated patches looks fine, but out of curiosity why did you opt for using `unlikely()` twice instead of wrapping then entire conditional?

Reply:
> That's a little troubling, but as I understand it you're no longer able to reproduce this in latter -rc's?

Yes, I didn't see it in -rc2 or after, but I could have been just lucky. Also I only looked through block/. It could have been mm/ fix if any.

I don't mind dropping this entire PR though. It'll eventually fail with or without this conditional if bio corruption happens. We just don't get to see bunch of kernel stack with this change.

> why did you opt for using unlikely() twice instead of wrapping then entire conditional?

I probably should have used single `unlikely()`. I thought gcc might generate the same code, but will fix it to use 1 unlikely.

Reply:
Updated + rebased.

Feel free to close this if you don't want this change given 
> I don't mind dropping this entire PR though. It'll eventually fail with or without this conditional if bio corruption happens. We just don't get to see bunch of kernel stack with this change.